spec.html

```diff
@@ -37627,29 +37597,55 @@ <h1>Array.prototype.findLast ( _predicate_ [ , _thisArg_ ] )</h1>
+          <p>This method calls _predicate_ once for each element of the array, in descending index order, until it finds one where _predicate_ returns a value that coerces to *true*. If such an element is found, `findLastIndex` immediately returns the index of that element value. Otherwise, `findLastIndex` returns -1.</p>
+          <p>See FindViaPredicate for additional information.</p>
         </emu-note>
         <p>This method performs the following steps when called:</p>
         <emu-alg>
           1. Let _O_ be ? ToObject(*this* value).
           1. Let _len_ be ? LengthOfArrayLike(_O_).
-          1. If IsCallable(_predicate_) is *false*, throw a *TypeError* exception.
-          1. Let _k_ be _len_ - 1.
-          1. Repeat, while _k_ ≥ 0,
-            1. Let _Pk_ be ! ToString(𝔽(_k_)).
-            1. Let _kValue_ be ? Get(_O_, _Pk_).
-            1. Let _testResult_ be ToBoolean(? Call(_predicate_, _thisArg_, « _kValue_, 𝔽(_k_), _O_ »)).
-            1. If _testResult_ is *true*, return 𝔽(_k_).
-            1. Set _k_ to _k_ - 1.
-          1. Return *-1*<sub>𝔽</sub>.
+          1. Let _findRec_ be ? FindViaPredicate(_O_, _len_, ~descending~, _predicate_, _thisArg_).
+          1. Return _findRec_.[[Index]].
         </emu-alg>
         <emu-note>
           <p>This method is intentionally generic; it does not require that its *this* value be an Array object. Therefore it can be transferred to other kinds of objects for use as a method.</p>
         </emu-note>
+
+        <emu-clause id="sec-findviapredicate" type="abstract operation">
+          <h1>
+            FindViaPredicate (
+              _O_: an Object,
+              _len_: a non-negative integer,
+              _direction_: ~ascending~ or ~descending~,
+              _predicate_: an ECMAScript language value,
+              _thisArg_: an ECMAScript language value,
+            ): either a normal completion containing a Record with fields [[Index]] (an integral Number) and [[Value]] (an ECMAScript language value) or a throw completion
+          </h1>
+          <dl class="header">
+            <dt>description</dt>
+            <dd>
+              <p>_O_ should be an array-like object or a TypedArray. This operation calls _predicate_ once for each element of _O_, in either ascending index order or descending index order (as indicated by _direction_), until it finds one where _predicate_ returns a value that coerces to *true*. At that point, this operation returns a Record that gives the index and value of the element found. If no such element is found, this operation returns a Record that specifies *-1*<sub>𝔽</sub> for the index and *undefined* for the value.</p>
```

Comment:
```suggestion
              <p>_O_ should be an array-like object or a TypedArray. This operation calls _predicate_ once for each index of of _O_ less than _len_, in either ascending index order or descending index order (as indicated by _direction_), until it finds one where _predicate_ returns a value that coerces to *true*. At that point, this operation returns a Record that gives the index and value of the element found. If no such element is found, this operation returns a Record that specifies *-1*<sub>𝔽</sub> for the index and *undefined* for the value.</p>
```

Reply:
Unlike above, the "once for each index" wording doesn't bother me here, because the very next paragraph says what's passed to `_predicate_`.

I think I'm slightly positive on this change, as it better conveys FindViaPredicate's approach to 'holes'.